src/target/llvm/codegen_hexagon.cc

```diff
@@ -0,0 +1,812 @@
+    } else {
+      CHECK_EQ(buf->getType(), t_tvm_array_->getPointerTo());
+    }
+    /* The following "kinds" are accessing the members of DLTensor:
+       typedef struct {
+         void* data;            kArrData
+         DLContext ctx;         kArrDeviceType (ctx.device_type)
+                                kArrDeviceId (ctx.device_id)
+         int ndim;              kArrNDim
+         DLDataType dtype;      kArrTypeCode (dtype.code)
+                                kArrTypeBits (dtype.bits)
+                                kArrTypeLanes (dtype.lanes)
+         int64_t* shape;        kArrShape
+         int64_t* strides;      kArrStrides
+         uint64_t byte_offset;  kArrByteOffset
+       } DLTensor;
+    */
+    llvm::Value* base_gep = builder_->CreateInBoundsGEP(buf, index, "base_gep");
+    if (kind == builtin::kArrAddr) {
+      return base_gep;
+    }
+    llvm::Value* field_gep = builder_->CreateInBoundsGEP(
+        base_gep, {ConstInt32(0), ConstInt32(field_index.at(kind))}, "field_gep");
+    switch (kind) {
+      // These fields have no sub-fields.
+      case builtin::kArrData:
+      case builtin::kArrNDim:
+      case builtin::kArrShape:
+      case builtin::kArrStrides:
+      case builtin::kArrByteOffset:
+        return field_gep;
+    }
+    return builder_->CreateInBoundsGEP(
+        field_gep, {ConstInt32(0), ConstInt32(subfield_index.at(kind))}, "subfield_gep");
+  }
+
+  if (kind == builtin::kTVMValueContent) {
+    /* TVMValue is a union:
+       typedef union {
+         int64_t v_int64;
+         double v_float64;
+         void* v_handle;
+         const char* v_str;
+         TVMType v_type;
+         TVMContext v_ctx;
+       } TVMValue;
+    */
+    CHECK_EQ(t.lanes(), 1);
+    CHECK(t.is_handle() || t.bits() == 64);
+    if (t.is_int()) {
+      buf = builder_->CreatePointerCast(buf, t_int64_->getPointerTo());
+      return builder_->CreateInBoundsGEP(buf, index);
+    } else if (t.is_float()) {
+      buf = builder_->CreatePointerCast(buf, t_float64_->getPointerTo());
+      return builder_->CreateInBoundsGEP(buf, index);
+    } else {
+      CHECK(t.is_handle());
+      buf = builder_->CreatePointerCast(buf, t_tvm_value_->getPointerTo());
+      buf = builder_->CreateInBoundsGEP(buf, index);
+      return builder_->CreatePointerCast(buf, t_void_p_->getPointerTo());
+    }
+  }
+
+  assert(!"Unknown kind");
+  return nullptr;
+}
+
+namespace {
+// Check if the function matches the TVMBackendPackedCFunc prototype.
+bool UsesExportABI(const PrimFunc& f) {
+  if (f->attrs.defined()) {
+    auto it = f->attrs->dict.find("calling_conv");
+    return it != f->attrs->dict.end() &&
+           Downcast<Integer>((*it).second) == CallingConv::kCPackedFunc;
+  }
+  return false;
+}
+
+__attribute__((unused)) std::ostream& operator<<(std::ostream& os, const llvm::Module& m) {
+  std::string ms;
+  llvm::raw_string_ostream sos(ms);
+  sos << m;
+  os << sos.str();
+  return os;
+}
```

Comment:
Which condition will we use it?

Reply:
There is a diagnostic code at [line 729](https://github.com/apache/incubator-tvm/pull/6261/files#diff-ed6a1469ccdc4ff4ac97357c972035dfR728-R729) that uses this.  It's commented out, but we find it very useful, so I kept it in the code.  Maybe it should be enabled with an environment variable instead, I'm not sure if TVM has an established convention for things like this.

Reply:
As far as i know, tvm doesn't have.